server/test/services/program/predicate/PredicateGeneratorTest.java

```diff
@@ -0,0 +1,316 @@
+
+    assertThat(predicateDefinition.predicateFormat().get())
+        .isEqualTo(PredicateDefinition.PredicateFormat.SINGLE_QUESTION);
+    assertThat(predicateDefinition.action()).isEqualTo(PredicateAction.HIDE_BLOCK);
+    assertThat(predicateDefinition.getQuestions())
+        .isEqualTo(ImmutableList.of(testQuestionBank.applicantAddress().id));
+    assertThat(predicateDefinition.rootNode())
+        .isEqualTo(
+            PredicateExpressionNode.create(
+                LeafOperationExpressionNode.builder()
+                    .setQuestionId(testQuestionBank.applicantAddress().id)
+                    .setScalar(Scalar.ZIP)
+                    .setOperator(Operator.EQUAL_TO)
+                    .setComparedValue(PredicateValue.of("98144"))
+                    .build()));
+  }
+
+  @Test
+  public void generatePredicateDefinition_multiQuestion_multiValue() throws Exception {
+    DynamicForm form =
+        buildForm(
+            ImmutableMap.of(
+                "predicateAction",
+                "SHOW_BLOCK",
+                String.format("question-%d-scalar", testQuestionBank.applicantAddress().id),
+                "ZIP",
+                String.format("question-%d-operator", testQuestionBank.applicantAddress().id),
+                "EQUAL_TO",
+                String.format(
+                    "group-1-question-%d-predicateValue", testQuestionBank.applicantAddress().id),
+                "98144",
+                String.format(
+                    "group-2-question-%d-predicateValue", testQuestionBank.applicantAddress().id),
+                "10001",
+                String.format("question-%d-scalar", testQuestionBank.applicantDate().id),
+                "DATE",
+                String.format("question-%d-operator", testQuestionBank.applicantDate().id),
+                "EQUAL_TO",
+                String.format(
+                    "group-1-question-%d-predicateValue", testQuestionBank.applicantDate().id),
+                "2023-01-01",
+                String.format(
+                    "group-2-question-%d-predicateValue", testQuestionBank.applicantDate().id),
+                "2023-02-02"));
+
+    PredicateDefinition predicateDefinition =
+        predicateGenerator.generatePredicateDefinition(form, readOnlyQuestionService);
+
+    assertThat(predicateDefinition.predicateFormat().get())
+        .isEqualTo(PredicateDefinition.PredicateFormat.OR_OF_SINGLE_LAYER_ANDS);
+    assertThat(predicateDefinition.action()).isEqualTo(PredicateAction.SHOW_BLOCK);
+    assertThat(predicateDefinition.getQuestions())
+        .containsExactlyInAnyOrder(
+            testQuestionBank.applicantAddress().id, testQuestionBank.applicantDate().id);
+
+    assertThat(predicateDefinition.rootNode().getType()).isEqualTo(PredicateExpressionNodeType.OR);
+    assertThat(predicateDefinition.rootNode().getOrNode().children())
+        .containsExactlyInAnyOrder(
```

Comment:
This needs to be in the exact order I believe so that we can redisplay them in the correct order.

Reply:
discussed offline

Reply:
The or nodes need to be in order as it's the rows of data, so this shouldn't be "AnyOrder" right?

Reply:
! thank you

changed assertion to `containsExactly`